### DISCUSSION TITLE
Auto-resizing panes by dragging divider

### DISCUSSION THREAD

**Discussion by cardinalgeo** [Kitware/trame#291] [Q&A] [answered]
Hello Trame team, 

I'm interested in subdividing a trame dashboard into two panes that automatically resize upon dragging the divider that vertically separates them. An example Vue.js library that accomplishes this is `splitpanes` ([here](https://antoniandre.github.io/splitpanes/)). I've played with `trame-grid-layout`, but its underlying Vue.js library doesn't seem to include this sort of auto-resizing (only manual resizing and dragging-induced translation of panes). Would you have any suggestions for how I might proceed in order to incorporate this functionality? 

Thanks, Robert

**Comment by jourdain** [comment, marked as answer]
You could create a repo and start with the trame-cookiecutter for just a component. 
Otherwise, if it was a vue plugin you could have used it with something like below. 

```python
from trame.app import get_server
from trame.ui.vuetify import VAppLayout
from trame.widgets import vuetify

server = get_server()

split_pane_module = {
    "scripts": ["https://unpkg.com/splitpanes@legacy"],
    "styles": ["https://unpkg.com/splitpanes@legacy/dist/splitpanes.css"],
    "vue_use": ["xxx"] # FIXME
}
server.enable_module(split_pane_module)

# Since splitpanes is not a vue plugin we need to wrap it...
# We are missing a JS object that does the following
#
# window.xxx = {
#     install(vue) {
#         Object.keys(window.splitpanes).forEach((name) => {
#             vue.component(name, window.splitpanes[name]);
#         });
#     }
# }

with VAppLayout(server) as layout:
    layout.root.add_child("""
    <splitpanes class="default-theme">
        <pane min-size="20">
            1
            <br>
            <em class="specs"> I have a min width of 20%<em>
        </pane>
        <pane>
            <splitpanes class="default-theme" horizontal>
                <pane min-size="15">
                    2
                    <br>
                    <em class="specs"> I have a min height of 15%</em>
                </pane>
                <pane>3</pane>
                <pane>4</pane>
            </splitpanes>
        </pane>
        <pane>
            5
        </pane>
    </splitpanes>
    """)

server.start()
```

You can look at that [repo](https://github.com/Kitware/trame-xterm) as an example except that you don't need to write the components. You just need to get them from npm and expose them within the [main](https://github.com/Kitware/trame-xterm/blob/master/vue-components/src/main.js).

Another option is to have no JS build and just bundle your own JS-install-function along with the 2 other files that are part of splitpanes. An example of it can be found in [trame-vtk](https://github.com/Kitware/trame-vtk) except that the JS code is actually a vue plugin (so no extra js file to add).

HTH

**Comment by jourdain** [comment]
Actually I made a working example [here](https://github.com/Kitware/trame/blob/master/examples/validation/discussion/291.py).

**Comment by cardinalgeo** [comment]
Hi Seb, thank you! I got your example to work, and I appreciate the time you took to make it. I have a follow-up question; I'd of course like to place more than just text in each pane. In the case of the `RemoteSelection` example found [here](https://github.com/Kitware/trame/blob/master/examples/06_vtk/Applications/RemoteSelection/app.py) in the `trame` repo, I'd like to place the `vtkRemoteView` object in one pane and the plotly `Figure` object in the other. Would you have any pointers for how I might go about doing this? How would I embed each object in the html added to `layout.root` in the `splitpanes` example you gave? 

if the question is quite simple, my apologies — my javascript/html experience is a bit limited. If answering it is fairly basic but would take too much time, I'd appreciate if you could point me in the right direction to figure it out!

**Comment by jourdain** [comment]
Just make 2 Python classes for Splipanes and Pane using the same model that we do in every widgets (i.e. [leaflet](https://github.com/Kitware/trame-leaflet/blob/master/trame_leaflet/widgets/leaflet.py#L40-L69)) where you just list the property and event mapping that you want to achieve between Python and Vue.

**Comment by jourdain** [comment]
```python
from trame_client.widgets.core import AbstractElement

class Splipanes(AbstractElement):
    def __init__(self, children=None, **kargs):
         super().__init__("splitpanes", children, **kwargs)
         self._attr_names += [
            # ...
         ]
         self._event_names += [
            # ...
         ]


class Pane(AbstractElement):
    def __init__(self, children=None, **kargs):
         super().__init__("pane", children, **kwargs)
         self._attr_names += [
            ("min_size", "min-size"),
             # ...
         ]
         self._event_names += [
             # ...
         ]
```

**Reply by cardinalgeo** [reply]
This was a big help, thanks Seb! I ultimately created the subclass `HtmlElement` and subsequently subclassed that for both `Splitpanes` and `Pane`, analogous to how it's done in `trame-leaflet`. I appreciate the examples!

**Reply by jourdain** [reply]
As you can see it is not bad at all to integrate anything that is available out there...

**Reply by cardinalgeo** [reply]
Yeah, after you connected the dots for me this became pretty clear — it has stirred up some interesting ideas to explore!

**Reply by jourdain** [reply]
Ideally, those external packages should be integrated into a python package (like we do for vtk, plotly, vuetify, ...) to streamline usage and integration.
